tdct/deconvolution.py

```diff
@@ -0,0 +1,864 @@
+    #Internal class to handle progress bar
+
+    ''' This code is not giving correct results
+    This is probably because the usage of the psf flipped
+    In a later version doRLDeconvolution5 (based on DeconvolutionLab2 instead of using psf-flipped in the 2nd convolution,
+    a correlation is performed. '''
+
+    #Estimate progress iterations
+    nProgrIter = 2*niter + 6 #Check if ok
+    progr0 = _progrBarHandle(qtprocessbar, nProgrIter) #Sets up
+
+    #Read data
+    if os.path.isfile(datapath) is True and os.path.isfile(psfdatapath) is True and niter>0:
+        if debug is True: print(clrmsg.DEBUG, "Loading images: ", datapath," , ",psfdatapath)
+        progr0.increment() #1
+
+        data_np = tf.imread(datapath)
+        psf_np = tf.imread(psfdatapath)
+
+        progr0.increment() #2
+
+        #Normalise both data (and also converts)
+
+        data_np_norm = _convertAndNormalise(data_np)
+        psf_np_norm = _convertAndNormalise(psf_np)
+
+        progr0.increment() #3
+
+        #Code inspired by scikit-image Richardson-lucy
+        #https://github.com/scikit-image/scikit-image/blob/main/skimage/restoration/deconvolution.py#L383-L443
+        psf_t = np.flip(psf_np_norm)
+        xn1 = np.full(data_np_norm.shape,0.5, dtype=data_np_norm.dtype)
+
+        progr0.increment() #4
+        
+        if debug is True: print(clrmsg.DEBUG, "Begining iterative deconvolution niter=", niter)
+
+        for i in range(niter):
+            xn=xn1
+            
+            Hx = scipy.signal.convolve(xn,psf_np_norm, mode='same')
+            yhx = np.divide(data_np_norm,Hx)
+
+            progr0.increment()
+
+            htyhx = scipy.signal.convolve(yhx,psf_t, mode='same')
+            xn1 = np.multiply(xn, htyhx)
+
+            progr0.increment()
+
+        #At the end xn1 should have the deconvoluted data
+        #Convert to uint8
+        data_deconv_norm_256 = _convertAndNormalise(xn1)*256
+        data_deconv_uint8 = data_deconv_norm_256.astype('uint8')
+        
+        progr0.increment() #5
+
+        #Save data
+        fpath,fname = os.path.split(datapath)
+        fname0 = os.path.join(fpath, "DeconvRL"+str(niter)+"it_"+fname)
+        tf.imsave(fname0 ,data_deconv_uint8)
+
+        progr0.setmax()
+ 
+        if debug is True: print(clrmsg.DEBUG, "Completed deconvolution, file saved: ", fname0)
+
+def doRLDeconvolution2(datapath , psfdatapath , niter=0, qtprocessbar=None):
+    '''Deconvolute data using the Richardson Lucy algorithm.
+    This code is based on the source code for DeconvolutionLab2
+    https://github.com/Biomedical-Imaging-Group/DeconvolutionLab2/blob/master/src/main/java/deconvolution/algorithm/RichardsonLucy.java
+    
+    This is not working well, beads are being duplicated in the final volume'''
+
+    #From https://github.com/scipy/scipy/blob/803e52d7e82cfc027daa55426466da29bc303b5c/scipy/signal/signaltools.py#L385
+    def _centered(arr, newshape):
```

Comment:
Internal function probably doesn't need to be named with a "_" at the start because it is already in the local scope so making it as "hidden" doesn't really serve a purpose.